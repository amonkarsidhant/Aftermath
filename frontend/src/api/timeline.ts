export interface TimelineEvent {
  source: string;
  timestamp: string;
  description: string;
}

export async function fetchTimelineEvents(
  start?: Date,
  end?: Date,
  providers?: string[]
): Promise<TimelineEvent[]> {
  const params = new URLSearchParams();
  if (start) params.set('start', start.toISOString());
  if (end) params.set('end', end.toISOString());
  if (providers && providers.length) params.set('providers', providers.join(','));

  const res = await fetch(`/timeline?${params.toString()}`);
  if (!res.ok) {
    throw new Error('Failed to fetch timeline events');
  }
  const data = await res.json();
  return data.events as TimelineEvent[];
}
