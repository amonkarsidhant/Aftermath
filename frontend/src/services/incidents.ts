import type { Incident } from '../types';
import { config } from '../utils/config';
import mockIncidents from '../utils/mock/incidents.json';

export async function fetchIncidents(): Promise<Incident[]> {
  if (config.useMockData) {
    return mockIncidents as Incident[];
  }
  const res = await fetch('/incidents');
  if (!res.ok) {
    throw new Error('failed to fetch incidents');
  }
  const data = await res.json();
  return data.incidents as Incident[];
}
