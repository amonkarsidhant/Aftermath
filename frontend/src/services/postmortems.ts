import type { Postmortem } from '../types';
import { config } from '../utils/config';
import mockPostmortems from '../utils/mock/postmortems.json';
import { fetchWithShare } from './share';

export async function searchPostmortems(q: string): Promise<Postmortem[]> {
  if (config.useMockData) {
    const normalized = q.toLowerCase();
    return (mockPostmortems as Postmortem[]).filter(
      (pm) =>
        pm.title.toLowerCase().includes(normalized) ||
        pm.summary.toLowerCase().includes(normalized) ||
        pm.tags.some((t) => t.toLowerCase().includes(normalized))
    );
  }
  const res = await fetchWithShare(
    `/postmortems/search?q=${encodeURIComponent(q)}`
  );
  if (!res.ok) throw new Error('failed to search');
  const data = await res.json();
  return data.postmortems as Postmortem[];
}
