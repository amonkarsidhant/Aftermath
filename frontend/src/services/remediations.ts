export interface AgendaNotes {
  well: string;
  wrong: string;
  lucky: string;
}

export interface Remediation {
  id: string;
  provider: string;
  description: string;
  url: string;
}

const AGENDA_KEY = 'review_agenda';
const ACTIONS_KEY = 'remediation_actions';

export function loadAgenda(): AgendaNotes {
  const raw = localStorage.getItem(AGENDA_KEY);
  return raw ? JSON.parse(raw) : { well: '', wrong: '', lucky: '' };
}

export function saveAgenda(notes: AgendaNotes): void {
  localStorage.setItem(AGENDA_KEY, JSON.stringify(notes));
}

export function loadActions(): Remediation[] {
  const raw = localStorage.getItem(ACTIONS_KEY);
  return raw ? JSON.parse(raw) : [];
}

function saveActions(actions: Remediation[]): void {
  localStorage.setItem(ACTIONS_KEY, JSON.stringify(actions));
}

export async function createAction(
  description: string,
  provider: 'jira' | 'servicenow'
): Promise<Remediation> {
  const res = await fetch('/api/actions', {
    method: 'POST',
    headers: { 'Content-Type': 'application/json' },
    body: JSON.stringify({ description, provider }),
  });
  const data = await res.json();
  const action: Remediation = {
    id: data.id,
    url: data.url,
    description,
    provider,
  };
  const actions = loadActions();
  actions.push(action);
  saveActions(actions);
  return action;
}
