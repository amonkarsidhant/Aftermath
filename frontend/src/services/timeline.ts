export interface TimelineEvent {
  source: string;
  timestamp: string;
  description: string;
}

import { config } from '../utils/config';
import mockEvents from '../utils/mock/timeline.json';

export async function fetchTimelineEvents(
  start?: Date,
  end?: Date,
  providers?: string[]
): Promise<TimelineEvent[]> {
  if (config.useMockData) {
    let events = mockEvents as TimelineEvent[];
    if (start) events = events.filter((e) => new Date(e.timestamp) >= start);
    if (end) events = events.filter((e) => new Date(e.timestamp) <= end);
    if (providers && providers.length) {
      events = events.filter((e) => providers.includes(e.source));
    }
    return events;
  }

  const params = new URLSearchParams();
  if (start) params.set('start', start.toISOString());
  if (end) params.set('end', end.toISOString());
  if (providers && providers.length) params.set('providers', providers.join(','));

  const res = await fetch(`/timeline?${params.toString()}`);
  if (!res.ok) {
    let message = 'Failed to fetch timeline events';
    try {
      const err = await res.json();
      message = err.error || err.message || message;
    } catch {
      try {
        const text = await res.text();
        if (text) message = text;
      } catch {
        /* ignore */
      }
    }
    throw new Error(message);
  }
  const data = await res.json();
  return data.events as TimelineEvent[];
}
