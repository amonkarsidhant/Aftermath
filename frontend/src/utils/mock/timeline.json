[
  {
    "source": "monitoring",
    "timestamp": "2024-01-01T00:00:00Z",
    "description": "Alert triggered",
    "category": "system"
  },
  {
    "source": "incident",
    "timestamp": "2024-01-01T00:05:00Z",
    "description": "Investigation started",
    "category": "human"
  },
  {
    "source": "resolution",
    "timestamp": "2024-01-01T00:30:00Z",
    "description": "Issue resolved",
    "category": "system"
  }
]
