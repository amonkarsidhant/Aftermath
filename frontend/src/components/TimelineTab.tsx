import { useEffect, useState } from 'react';
import { fetchTimelineEvents, TimelineEvent } from '../api/timeline';

export default function TimelineTab() {
  const [events, setEvents] = useState<TimelineEvent[]>([]);
  const [start, setStart] = useState('');
  const [end, setEnd] = useState('');
  const [selectedProviders, setSelectedProviders] = useState<string[]>([]);
  const [error, setError] = useState<string | null>(null);

  useEffect(() => {
    async function load() {
      try {
        const s = start ? new Date(start) : undefined;
        const e = end ? new Date(end) : undefined;
        const data = await fetchTimelineEvents(s, e);
        setEvents(data);
        const providers = Array.from(new Set(data.map((ev) => ev.source)));
        setSelectedProviders(providers);
        setError(null);
      } catch (err) {
        setError(err instanceof Error ? err.message : String(err));
      }
    }
    load();
  }, [start, end]);

  const toggleProvider = (p: string) => {
    setSelectedProviders((prev) =>
      prev.includes(p) ? prev.filter((x) => x !== p) : [...prev, p]
    );
  };

  const filtered = events.filter((ev) => {
    const time = new Date(ev.timestamp).getTime();
    const s = start ? new Date(start).getTime() : -Infinity;
    const e = end ? new Date(end).getTime() : Infinity;
    return (
      selectedProviders.includes(ev.source) &&
      time >= s &&
      time <= e
    );
  });

  const minTime = Math.min(...filtered.map((e) => new Date(e.timestamp).getTime()));
  const maxTime = Math.max(...filtered.map((e) => new Date(e.timestamp).getTime()));
  const height = 300;
  const getY = (time: number) => {
    if (!isFinite(minTime) || !isFinite(maxTime) || minTime === maxTime) {
      return height / 2;
    }
    return ((time - minTime) / (maxTime - minTime)) * height;
  };

  const providers = Array.from(new Set(events.map((e) => e.source)));

  return (
    <div className="space-y-4">
      {error && (
        <div role="alert" className="p-2 bg-red-100 text-red-600 rounded">
          {error}
        </div>
      )}
      <div className="flex flex-wrap gap-4 items-end">
        <div>
          <label className="block text-sm">Start</label>
          <input
            type="date"
            value={start}
            onChange={(e) => setStart(e.target.value)}
            className="border p-1 rounded"
          />
        </div>
        <div>
          <label className="block text-sm">End</label>
          <input
            type="date"
            value={end}
            onChange={(e) => setEnd(e.target.value)}
            className="border p-1 rounded"
          />
        </div>
        <div className="flex gap-2">
          {providers.map((p) => (
            <label key={p} className="flex items-center gap-1">
              <input
                type="checkbox"
                checked={selectedProviders.includes(p)}
                onChange={() => toggleProvider(p)}
              />
              {p}
            </label>
          ))}
        </div>
      </div>
      <div className="overflow-x-auto bg-white dark:bg-neutral-800 border border-neutral-200 dark:border-neutral-700 rounded p-4">
        <svg height={height} className="w-full">
          <line x1={50} y1={0} x2={50} y2={height} stroke="currentColor" />
          {filtered.map((ev, idx) => {
            const y = getY(new Date(ev.timestamp).getTime());
            return (
              <g key={idx} transform={`translate(50, ${y})`}>
                <circle r={5} fill="currentColor" />
                <text x={10} y={5} className="text-sm">
                  {`${new Date(ev.timestamp).toLocaleString()} - ${ev.source}: ${ev.description}`}
                </text>
              </g>
            );
          })}
        </svg>
      </div>
    </div>
  );
}
