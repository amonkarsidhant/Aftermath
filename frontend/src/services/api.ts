import { useQuery } from '@tanstack/react-query';
import type { Incident, Postmortem } from '../types';
import { fetchIncidents } from './incidents';
import { searchPostmortems } from './postmortems';

export function useIncidents() {
  return useQuery<Incident[]>({
    queryKey: ['incidents'],
    queryFn: fetchIncidents,
  });
}

export function usePostmortems(
  query: string,
  options?: { enabled?: boolean }
) {
  return useQuery<Postmortem[]>({
    queryKey: ['postmortems', query],
    queryFn: () => searchPostmortems(query),
    ...options,
  });
}
