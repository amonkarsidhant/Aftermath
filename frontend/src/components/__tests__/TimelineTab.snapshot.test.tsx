import { render, screen } from '@testing-library/react';
import TimelineTab from '../TimelineTab';
import { fetchTimelineEvents } from '../../services/timeline';

jest.mock('../../services/timeline');

describe('TimelineTab snapshot', () => {
  it('matches snapshot', async () => {
    (fetchTimelineEvents as jest.Mock).mockResolvedValue([
      {
        source: 'slack',
        timestamp: '2024-01-01T00:00:00Z',
        description: 'Slack event',
        category: 'human',
      },
      {
        source: 'pagerduty',
        timestamp: '2024-01-01T01:00:00Z',
        description: 'PagerDuty event',
        category: 'system',
      },
    ]);

    const { container } = render(<TimelineTab />);

    expect(await screen.findByText(/Slack event/)).toBeInTheDocument();
    expect(container).toMatchSnapshot();
  });
});
