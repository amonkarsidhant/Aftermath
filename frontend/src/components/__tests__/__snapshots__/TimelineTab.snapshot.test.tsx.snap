exports[`TimelineTab snapshot matches snapshot 1`] = `
<div class="space-y-4">
  <div class="flex flex-wrap gap-4 items-end">
    <div>
      <label class="block text-sm">Start</label>
      <input type="date" value="" class="border p-1 rounded" />
    </div>
    <div>
      <label class="block text-sm">End</label>
      <input type="date" value="" class="border p-1 rounded" />
    </div>
    <div class="flex gap-2">
      <label class="flex items-center gap-1">
        <input type="checkbox" checked />
        slack
      </label>
      <label class="flex items-center gap-1">
        <input type="checkbox" checked />
        pagerduty
      </label>
    </div>
    <div class="flex gap-2">
      <label class="flex items-center gap-1">
        <input type="checkbox" checked />
        Human
      </label>
      <label class="flex items-center gap-1">
        <input type="checkbox" checked />
        System
      </label>
    </div>
  </div>
  <div class="overflow-x-auto bg-white dark:bg-neutral-800 border border-neutral-200 dark:border-neutral-700 rounded p-4">
    <svg height="300" class="w-full">
      <line x1="50" y1="0" x2="50" y2="300" stroke="currentColor" />
      <g transform="translate(50, 0)">
        <circle r="5" fill="currentColor" />
        <text x="10" y="5" class="text-sm">1/1/2024, 12:00:00 AM - slack: Slack event</text>
      </g>
      <g transform="translate(50, 300)">
        <circle r="5" fill="currentColor" />
        <text x="10" y="5" class="text-sm">1/1/2024, 1:00:00 AM - pagerduty: PagerDuty event</text>
      </g>
    </svg>
  </div>
</div>
`;
